pkg/action/upgrade.go

```diff
@@ -306,11 +317,44 @@ func (u *Upgrade) performUpgrade(originalRelease, upgradedRelease *release.Relea
 	if err := u.cfg.Releases.Create(upgradedRelease); err != nil {
 		return nil, err
 	}
+	rChan := make(chan resultMessage)
+	go u.releasingUpgrade(rChan, upgradedRelease, current, target, originalRelease)
+	go u.handleSignals(rChan, upgradedRelease)
+	result := <-rChan
+
+	return result.r, result.e
+}
 
+// Function used to lock the Mutex, this is important for the case when the atomic flag is set.
+// In that case the upgrade will finish before the rollback is finished so it is necessary to wait for the rollback to finish.
+// The rollback will be trigger by the function failRelease
+func (u *Upgrade) reportToPerformUpgrade(c chan<- resultMessage, rel *release.Release, created kube.ResourceList, err error) {
+	u.Lock.Lock()
+	if err != nil {
+		rel, err = u.failRelease(rel, created, err)
+	}
+	c <- resultMessage{r: rel, e: err}
+	u.Lock.Unlock()
+}
+
+// Setup listener for SIGINT and SIGTERM
+func (u *Upgrade) handleSignals(c chan<- resultMessage, upgradedRelease *release.Release) {
```

Comment:
Because pkg/action is a public API, this could be construed as a breaking change as this introduces new behaviour to interact with the API.

There could be cases where a program uses Helm's go libraries to perform upgrades in a CI/CD pipeline, and they do not want the upgrade to be interrupted when a SIGTERM/SIGINT are captured.

This should either be part of `cmd/helm` or added as an opt-in flag.

Reply:
One point of note: contexts should not be stored in a structure. They should be passed in the function signature.

Unfortunately that means we may have to introduce a new `UpgradeWithContext` function to accommodate contexts.

From https://pkg.go.dev/context

> Do not store Contexts inside a struct type; instead, pass a Context explicitly to each function that needs it. The Context should be the first parameter, typically named ctx:

Reply:
Thanks for the tip, because I am not well knowledgeable about the context package I will start to dig in some docs and explanations, and in all the flow I try to add the context in the helm/cmd and see what we can do